### DISCUSSION TITLE
Won't be able to make tsfresh work on Windows

### DISCUSSION THREAD

**Discussion by mkatgithub123** [blue-yonder/tsfresh#976] [Q&A] [answered]
Trying to make tsfresh work under Windows - however, I can't manage to do so. I am using Python 3.10.6

**1. What I did**

Terminal
```
conda create --name tsfresh_test
conda activate tsfresh_test
conda install pip
pip install pandas matplotlib tsfresh
```

Then, after choosing the resulting Interpreter, I ran what basically is the example code provided on the `tsfresh` readthedocs homepage:

```
#%%
from tsfresh.examples.robot_execution_failures import load_robot_execution_failures
timeseries, y = load_robot_execution_failures()

print(timeseries.head())

#%%
import matplotlib.pyplot as plt
timeseries[timeseries['id'] == 3].plot(subplots=True, sharex=True, figsize=(10,10))
plt.show()

# %%
timeseries[timeseries['id'] == 20].plot(subplots=True, sharex=True, figsize=(10,10))
plt.show()

# %%
from tsfresh import extract_features
if __name__ == '__main__':
    extracted_features = extract_features(timeseries, column_id="id", column_sort="time")
# %%

```

**2. What I got**

The runs properly  until `extracted_features = extract_features(....)` and from thereon runs ad infinitum, calling itself over and over again FROM its beginning. It repeatedly creates the plots specified via `matplotlib`, but the `tsfresh` feature progress bar does not change, it stays at 0%.  Based on this, I suspect this is somehow related to multiprocessing (which I don't have any idea of), which is why I included the line `if __name__ == '__main__':` in my code which I found upon investigating problems others had with `tsfresh` and multiprocessing issues.

**3. What I tried**

I tried setting up the environment in different ways with different Python versions. To no avail. I also consider(ed) switching to an alternative package named `tsfel` which seems easier to access/more suitable for Windows users, but whose documentation is lacking (at least for me) a clear description of how to include custom features which I can't seem to make work either. The problem persists irrespective of whether I run the code in interactive mode or directly.

However, I am really in need of setting up a pipeline for extracting features from time series data in a structured and systematic fashion. 

Does anybody have a clue what could be done about this? Already spent quite some time on this :-/

**Comment by nils-braun** [comment, marked as answer]
Hi @mkatgithub123 - sorry for the massive delay in the answer. If you are using `if __name__ == '__main__':` you want to put everything except the imports under it, like
```python
from tsfresh.examples.robot_execution_failures import load_robot_execution_failures
import matplotlib.pyplot as plt
from tsfresh import extract_features

if __name__ == '__main__':
    timeseries, y = load_robot_execution_failures()
    print(timeseries.head())
    
    timeseries[timeseries['id'] == 3].plot(subplots=True, sharex=True, figsize=(10,10))
    plt.show()

    timeseries[timeseries['id'] == 20].plot(subplots=True, sharex=True, figsize=(10,10))
    plt.show()
    
    extracted_features = extract_features(timeseries, column_id="id", column_sort="time")
```

The `__main__` line prevents Python to redo all the computations on multiprocessing, but only if it is "guarded" by it.
In general, it is good practise to always put your actual code (except imports and function definitions) into the main guard.
